# FINMA Python Lab 05: Lists, Tuples, Dictionaries, and Sets

## Overview

In this lab, you'll practice:
- Creating and manipulating lists
- Working with tuples (immutable sequences)
- Using dictionaries for key-value data
- Understanding sets and their operations
- Using the `range()` function
- List comprehensions
- Choosing the right data structure

**Important:** Complete your work and have it manually checked by your instructor.

---

## Part 1: Working with Lists

Lists are **ordered, mutable** collections. They can store multiple items and allow duplicates.

### Creating and Accessing Lists

In [13]:
# Example: List of stock prices
stock_prices = [150.75, 153.30, 149.50, 155.00, 157.25]
print("Stock prices:", stock_prices)

# Accessing elements
print("First stock price:", stock_prices[0])
print("Last stock price:", stock_prices[-1])
print("Second and third prices:", stock_prices[1:4:2])

Stock prices: [150.75, 153.3, 149.5, 155.0, 157.25]
First stock price: 150.75
Last stock price: 157.25
Second and third prices: [153.3, 155.0]


### Modifying Lists

Lists are mutable, meaning you can change them after creation.

In [19]:
# Modifying elements
stock_prices = [150.75, 153.30, 149.50, 155.00, 157.25]
stock_prices[1] = 154.00
print("Modified stock prices:", stock_prices)

# Adding elements
stock_prices.append(160.00)
print("After appending:", stock_prices)

# Inserting at a specific position
stock_prices.insert(2, 148.00)
print("After inserting at position 0:", stock_prices)

# Removing elements
stock_prices.remove(149.50)
print("After removing 149.50:", stock_prices)

# Removing by index
removed_price = stock_prices.pop(2)
print(f"Removed {removed_price}, remaining:", stock_prices)

Modified stock prices: [150.75, 154.0, 149.5, 155.0, 157.25]
After appending: [150.75, 154.0, 149.5, 155.0, 157.25, 160.0]
After inserting at position 0: [150.75, 154.0, 148.0, 149.5, 155.0, 157.25, 160.0]
After removing 149.50: [150.75, 154.0, 148.0, 155.0, 157.25, 160.0]
Removed 148.0, remaining: [150.75, 154.0, 155.0, 157.25, 160.0]


### Common List Methods

In [20]:
# Common list operations
prices = [150.75, 153.30, 149.50, 155.00, 157.25]

print("Length:", len(prices))
print("Maximum price:", max(prices))
print("Minimum price:", min(prices))
print("Sum of prices:", sum(prices))
print("Average price:", sum(prices) / len(prices))

# Sorting
prices.sort()
print("Sorted ascending:", prices)

prices.sort(reverse=True)
print("Sorted descending:", prices)

# Reversing
prices.reverse()
print("Reversed:", prices)

Length: 5
Maximum price: 157.25
Minimum price: 149.5
Sum of prices: 765.8
Average price: 153.16
Sorted ascending: [149.5, 150.75, 153.3, 155.0, 157.25]
Sorted descending: [157.25, 155.0, 153.3, 150.75, 149.5]
Reversed: [149.5, 150.75, 153.3, 155.0, 157.25]


---

## Part 2: Using the range() Function

The `range()` function generates sequences of numbers, which is very useful for creating lists and loops.

### Basic range() Usage

In [21]:
# range() returns a range object, not a list
print("range(11):", range(11))

# Convert to list to see the values
print("list(range(11)):", list(range(11)))

# range(start, stop)
print("list(range(1, 11)):", list(range(1, 11)))

# range(start, stop, step)
print("Even numbers 0-10:", list(range(0, 11, 2)))
print("Odd numbers 1-10:", list(range(1, 11, 2)))
print("Counting backwards:", list(range(10, 0, -1)))

range(11): range(0, 11)
list(range(11)): [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
list(range(1, 11)): [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Even numbers 0-10: [0, 2, 4, 6, 8, 10]
Odd numbers 1-10: [1, 3, 5, 7, 9]
Counting backwards: [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]


### Exercise 1: Creating Lists with range()

Use the `range()` function to create the following lists:

**a)** All numbers from 0 to 10 (inclusive)

In [1]:
# Exercise 1a: Numbers 0 to 10
# Write your code here

my_range = range(11)
my_list = list(my_range)

print(my_list)


[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


**b)** All numbers from 0 to 19 (inclusive)

In [2]:
# Exercise 1b: Numbers 0 to 19
# Write your code here

my_range = range(20)
my_list = list(my_range)

print(my_list)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]


**c)** All even numbers from 0 to 30 (inclusive)

In [4]:
# Exercise 1c: Even numbers 0 to 30
# Write your code here


print(list(range(0,31,2)))

[0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30]


---

## Part 3: Tuples

Tuples are **ordered, immutable** collections. Once created, they cannot be changed.

### Why Use Tuples?
- **Immutability**: Data that shouldn't change (e.g., coordinates, configuration)
- **Performance**: Slightly faster than lists
- **Dictionary keys**: Tuples can be used as keys, lists cannot

### Creating and Using Tuples

In [5]:
# Example: Tuple of stock information (symbol, company, price)
stock_info = ("AAPL", "Apple Inc.", 150.75)
print("Stock information:", stock_info)

# Accessing elements (same as lists)
print("Stock symbol:", stock_info[0])
print("Company name:", stock_info[1])
print("Stock price:", stock_info[2])

# Unpacking tuples
symbol, company, price = stock_info
print(f"{company} ({symbol}): ${price}")

Stock information: ('AAPL', 'Apple Inc.', 150.75)
Stock symbol: AAPL
Company name: Apple Inc.
Stock price: 150.75
Apple Inc. (AAPL): $150.75


In [ ]:
# Tuples are immutable - this will cause an error!
# Uncomment to see the error:
# stock_info[2] = 151.00  # TypeError: 'tuple' object does not support item assignment

### Multiple Tuples in a List

In [6]:
# List of stock tuples
stocks = [
    ("AAPL", "Apple Inc.", 150.75),
    ("GOOGL", "Alphabet Inc.", 138.21),
    ("MSFT", "Microsoft Corp.", 378.91)
]

# Accessing and unpacking
for symbol, company, price in stocks:
    print(f"{symbol}: ${price:.2f}")

AAPL: $150.75
GOOGL: $138.21
MSFT: $378.91


---

## Part 4: Dictionaries

Dictionaries are **unordered, mutable** collections of key-value pairs.

### Creating and Using Dictionaries

In [7]:
# Dictionary of stock prices
stock_prices = {
    "AAPL": 150.75,
    "GOOGL": 2800.50,
    "MSFT": 299.00
}
print("Stock prices:", stock_prices)

# Accessing elements
print("Price of AAPL:", stock_prices["AAPL"])

# Safer access with .get()
print("Price of TSLA:", stock_prices.get("TSLA", "Not found"))

# Modifying elements
stock_prices["GOOGL"] = 2820.00
print("Modified stock prices:", stock_prices)

# Adding elements
stock_prices["AMZN"] = 3400.00
print("After adding AMZN:", stock_prices)

# Removing elements
del stock_prices["MSFT"]
print("After removing MSFT:", stock_prices)

Stock prices: {'AAPL': 150.75, 'GOOGL': 2800.5, 'MSFT': 299.0}
Price of AAPL: 150.75
Price of TSLA: Not found
Modified stock prices: {'AAPL': 150.75, 'GOOGL': 2820.0, 'MSFT': 299.0}
After adding AMZN: {'AAPL': 150.75, 'GOOGL': 2820.0, 'MSFT': 299.0, 'AMZN': 3400.0}
After removing MSFT: {'AAPL': 150.75, 'GOOGL': 2820.0, 'AMZN': 3400.0}


### Dictionary Methods

In [8]:
stock_prices = {"AAPL": 150.75, "GOOGL": 2800.50, "MSFT": 299.00}

# Get all keys
print("Keys:", list(stock_prices.keys()))

# Get all values
print("Values:", list(stock_prices.values()))

# Get all key-value pairs
print("Items:", list(stock_prices.items()))

# Check if key exists
print("Is AAPL in dictionary?", "AAPL" in stock_prices)
print("Is TSLA in dictionary?", "TSLA" in stock_prices)

Keys: ['AAPL', 'GOOGL', 'MSFT']
Values: [150.75, 2800.5, 299.0]
Items: [('AAPL', 150.75), ('GOOGL', 2800.5), ('MSFT', 299.0)]
Is AAPL in dictionary? True
Is TSLA in dictionary? False


### Iterating Over Dictionaries

In [9]:
stock_prices = {"AAPL": 150.75, "GOOGL": 2800.50, "MSFT": 299.00}

# Iterate over keys
print("Stock symbols:")
for symbol in stock_prices:
    print(f"  {symbol}")

# Iterate over values
print("\nStock prices:")
for price in stock_prices.values():
    print(f"  ${price:.2f}")

# Iterate over key-value pairs
print("\nStock information:")
for symbol, price in stock_prices.items():
    print(f"  {symbol}: ${price:.2f}")

Stock symbols:
  AAPL
  GOOGL
  MSFT

Stock prices:
  $150.75
  $2800.50
  $299.00

Stock information:
  AAPL: $150.75
  GOOGL: $2800.50
  MSFT: $299.00


---

## Part 5: Sets

Sets are **unordered, mutable** collections of **unique** items. No duplicates allowed!

### Creating and Using Sets

In [10]:
# Set of stock symbols
stock_symbols = {"AAPL", "GOOGL", "MSFT"}
print("Stock symbols:", stock_symbols)

# Adding elements
stock_symbols.add("AMZN")
print("After adding AMZN:", stock_symbols)

# Adding a duplicate (no effect)
stock_symbols.add("AAPL")
print("After adding AAPL again:", stock_symbols)

# Removing elements
stock_symbols.remove("MSFT")
print("After removing MSFT:", stock_symbols)

Stock symbols: {'GOOGL', 'MSFT', 'AAPL'}
After adding AMZN: {'GOOGL', 'AMZN', 'MSFT', 'AAPL'}
After adding AAPL again: {'GOOGL', 'AMZN', 'MSFT', 'AAPL'}
After removing MSFT: {'GOOGL', 'AMZN', 'AAPL'}


### Set Operations

Sets are powerful for mathematical operations like union, intersection, and difference.

In [11]:
# Portfolio 1 and Portfolio 2
portfolio1 = {"AAPL", "GOOGL", "MSFT"}
portfolio2 = {"AMZN", "GOOGL", "TSLA", "MSFT"}

# Union: All unique stocks from both portfolios
all_stocks = portfolio1.union(portfolio2)
print("All unique stocks:", all_stocks)

# Intersection: Stocks in both portfolios
common_stocks = portfolio1.intersection(portfolio2)
print("Common stocks:", common_stocks)

# Difference: Stocks in portfolio1 but not in portfolio2
unique_to_p1 = portfolio1.difference(portfolio2)
print("Unique to portfolio 1:", unique_to_p1)

# Symmetric difference: Stocks in either but not both
exclusive_stocks = portfolio1.symmetric_difference(portfolio2)
print("Exclusive stocks:", exclusive_stocks)

All unique stocks: {'GOOGL', 'MSFT', 'TSLA', 'AMZN', 'AAPL'}
Common stocks: {'GOOGL', 'MSFT'}
Unique to portfolio 1: {'AAPL'}
Exclusive stocks: {'TSLA', 'AMZN', 'AAPL'}


### When to Use Sets

Sets are useful for:
- Removing duplicates from a list
- Fast membership testing (`x in set`)
- Mathematical set operations

In [12]:
# Example: Finding unique stock symbols from transaction history
transactions = ["AAPL", "GOOGL", "MSFT", "AAPL", "TSLA", "GOOGL", "AAPL"]
unique_symbols = set(transactions)
print("Transaction list:", transactions)
print("Unique symbols traded:", unique_symbols)
print("Number of unique symbols:", len(unique_symbols))

Transaction list: ['AAPL', 'GOOGL', 'MSFT', 'AAPL', 'TSLA', 'GOOGL', 'AAPL']
Unique symbols traded: {'GOOGL', 'MSFT', 'TSLA', 'AAPL'}
Number of unique symbols: 4


---

## Part 6: List Comprehensions

List comprehensions provide a concise way to create lists.

### Basic Syntax

```python
new_list = [expression for item in iterable]
new_list = [expression for item in iterable if condition]
```

In [13]:
# Example: Square each number
numbers = [1, 2, 3, 4, 5]

# Traditional way
squares_traditional = []
for num in numbers:
    squares_traditional.append(num ** 2)
print("Traditional:", squares_traditional)

# List comprehension
squares_comprehension = [num ** 2 for num in numbers]
print("Comprehension:", squares_comprehension)

Traditional: [1, 4, 9, 16, 25]
Comprehension: [1, 4, 9, 16, 25]


In [14]:
# Example: Convert prices from USD to EUR (assuming 1 USD = 0.85 EUR)
prices_usd = [150.75, 153.30, 149.50, 155.00]
prices_eur = [price * 0.85 for price in prices_usd]
print("USD prices:", prices_usd)
print("EUR prices:", prices_eur)

USD prices: [150.75, 153.3, 149.5, 155.0]
EUR prices: [128.1375, 130.305, 127.075, 131.75]


In [15]:
# Example: Filter prices above 150
prices = [150.75, 153.30, 149.50, 155.00, 148.00]
high_prices = [price for price in prices if price > 150]
print("All prices:", prices)
print("Prices above 150:", high_prices)

All prices: [150.75, 153.3, 149.5, 155.0, 148.0]
Prices above 150: [150.75, 153.3, 155.0]


---

## Programming Exercises

### Exercise 2: Multiply List Elements

You have the following list of numbers: `[1, 2, 3, 4, 5, 6]`

Iterate over this list, printing out each value multiplied by 10.

Find **two different solutions**:
1. Using a regular for loop with print
2. Using a list comprehension to create a new list

**Expected output (Solution 1):**
```
10
20
30
40
50
60
```

**Expected output (Solution 2):**
```
[10, 20, 30, 40, 50, 60]
```

In [ ]:
# Exercise 2: Solution 1 - Using for loop
numbers = [1, 2, 3, 4, 5, 6]

# Write your code here

#solution 1
for num in numbers:
    print(num*10)

#solution 2
numbers_10 = [num * 10 for num in numbers]
print(numbers_10)

10
20
30
40
50
60
[10, 20, 30, 40, 50, 60]


In [ ]:
# Exercise 2: Solution 2 - Using list comprehension
numbers = [1, 2, 3, 4, 5, 6]

# Write your code here



### Exercise 3: Vector Sum

Write a program that reads two 3D vectors (with x, y, z components) and calculates their sum.

A vector is represented as having three components: x, y, and z.

To add two vectors, you add their corresponding components:
- sum.x = first.x + second.x
- sum.y = first.y + second.y
- sum.z = first.z + second.z

**Examples:**
```
Please enter the values of the first vector (x, y, z): 1 2 3
Please enter the values of the second vector (x, y, z): 4 5 6
The resulting sum vector is:
x: 5
y: 7
z: 9
```

```
Please enter the values of the first vector (x, y, z): -5 30 -12
Please enter the values of the second vector (x, y, z): 15 -31 -30
The resulting sum vector is:
x: 10
y: -1
z: -42
```

**Hint:** Use a dictionary to store each vector:
```python
vector1 = {'x': x1, 'y': y1, 'z': z1}
```

In [34]:
# Exercise 3: Vector Sum
# Write your code here

first_vector = input("Please enter the values of the first vector (x,y,z): ").split()
second_vector = input("Please enter the values of the second vector (x,y,x): ").split()

vector1 = {
    "x": first_vector[0],
    "y": first_vector[1],
    "z": first_vector[2]
}

vector2 = {
    "x": second_vector[0],
    "y": second_vector[1],
    "z": second_vector[2]

}




print(f"The resulting sum vector is x: {vector1['x']+vector2['x']}, y: {vector1['y']+vector2['y']}, z: {vector1['z']+vector2['z']}")



The resulting sum vector is x: 14, y: 25, z: 36


### Exercise 4: Stock Price Analysis

Given the following list of daily stock prices, calculate:
1. The highest price
2. The lowest price
3. The average price
4. How many days the price was above average
5. The price change from first to last day

**Stock prices:** `[150.75, 153.30, 149.50, 155.00, 157.25, 152.80, 154.50]`

**Expected output:**
```
Highest price: $157.25
Lowest price: $149.50
Average price: $153.30
Days above average: 4
Price change: +$3.75
```

In [41]:
# Exercise 4: Stock Price Analysis
stock_prices = [150.75, 153.30, 149.50, 155.00, 157.25, 152.80, 154.50]

# max price
print(f"The max stock price is {max(stock_prices)}")

print(f"The min stock price is {min(stock_prices)}")

print(f"The average stock price is {sum(stock_prices)/len(stock_prices)}")

print(f"The number of days above 153 is {len([price for price in stock_prices if price > 153])}")

print(f"The stock price change is {stock_prices[-1]-stock_prices[0]}")


The max stock price is 157.25
The min stock price is 149.5
The average stock price is 153.29999999999998
The number of days above 153 is 4
The stock price change is 3.75


### Exercise 5: Portfolio Manager

Create a portfolio management system using a dictionary.

**Requirements:**
1. Create a dictionary with stock symbols as keys and quantities as values
2. Add a function to add stocks to the portfolio
3. Add a function to remove stocks from the portfolio
4. Add a function to display all stocks and quantities
5. Add a function to show total number of different stocks

**Example usage:**
```python
portfolio = {}
add_stock(portfolio, "AAPL", 100)
add_stock(portfolio, "GOOGL", 50)
add_stock(portfolio, "AAPL", 25)  # Should add to existing
display_portfolio(portfolio)
```

**Expected output:**
```
Portfolio:
  AAPL: 125 shares
  GOOGL: 50 shares
Total stocks: 2
```

In [10]:
# Exercise 5: Portfolio Manager

def buy_stock(portfolio, symbol, quantity):
    if symbol in portfolio:
        portfolio [symbol] += quantity
    else: 
        portfolio[symbol] = quantity
    print(f"The amount of {symbol} in the portfolio is {portfolio[symbol]}")


def sell_stock(portfolio, symbol, quantity):
    if symbol in portfolio:
        portfolio[symbol] -= quantity
    else:
        portfolio[symbol] = - quantity 
    print(f"The amount of {symbol} in the portfolio is {portfolio[symbol]}")

def display_portfolio(portfolio):
    for symbol, quantity in portfolio.items():
        print(f"The amount of {symbol} held is {quantity}")

# Test your functions
portfolio = {}
portfolio2 = {}
buy_stock(portfolio, "AAPL", 100)
buy_stock(portfolio, "GOOGL", 50)
buy_stock(portfolio, "AAPL", 25)
display_portfolio(portfolio)

The amount of AAPL in the portfolio is 100
The amount of GOOGL in the portfolio is 50
The amount of AAPL in the portfolio is 125
The amount of AAPL held is 125
The amount of GOOGL held is 50


### Exercise 6: Find Common and Unique Holdings

Two investors have different portfolios. Use sets to find:
1. Stocks that both investors own (intersection)
2. All unique stocks across both portfolios (union)
3. Stocks that only Investor 1 owns (difference)
4. Stocks that only one investor owns (symmetric difference)

**Portfolios:**
- Investor 1: `["AAPL", "GOOGL", "MSFT", "TSLA", "NVDA"]`
- Investor 2: `["AAPL", "AMZN", "MSFT", "META", "NFLX"]`

**Expected output:**
```
Common stocks: {'AAPL', 'MSFT'}
All unique stocks: {'AAPL', 'GOOGL', 'MSFT', 'TSLA', 'NVDA', 'AMZN', 'META', 'NFLX'}
Only Investor 1: {'GOOGL', 'TSLA', 'NVDA'}
Exclusive to one: {'GOOGL', 'TSLA', 'NVDA', 'AMZN', 'META', 'NFLX'}
```

In [ ]:

# Exercise 6: Find Common and Unique Holdings
investor1 = {"AAPL", "GOOGL", "MSFT", "TSLA", "NVDA"}
investor2 = {"AAPL", "AMZN", "MSFT", "META", "NFLX"}

# Write your code here
#sets are built to do venn diagram type things
common_stocks = investor1.intersection(investor2)
all_unique_stocks = investor1.union(investor2)
unique_to_i1 = investor1.difference(investor2)
exclusive_stocks = investor1.symmetric_difference(investor2)



print("Common stocks: ",common_stocks)
print("All unique stocks: ",all_unique_stocks)
print("Only Investor 1: ", unique_to_i1)
print("Exclusive to one: ",exclusive_stocks)





Common stocks:  {'MSFT', 'AAPL'}
All unique stocks:  {'GOOGL', 'META', 'AAPL', 'AMZN', 'NFLX', 'NVDA', 'MSFT', 'TSLA'}
Only Investor 1:  {'GOOGL', 'TSLA', 'NVDA'}
Exclusive to one:  {'GOOGL', 'META', 'TSLA', 'AMZN', 'NFLX', 'NVDA'}


### Exercise 7: Stock Information Database

Create a database of stock information using a dictionary where:
- Keys are stock symbols
- Values are tuples containing (company_name, sector, price)

**Requirements:**
1. Create the database with at least 5 stocks
2. Print all stocks in the Technology sector
3. Find the most expensive stock
4. Calculate the average price of all stocks

**Example database:**
```python
{
    "AAPL": ("Apple Inc.", "Technology", 150.75),
    "JPM": ("JPMorgan Chase", "Finance", 145.30),
    ...
}
```

In [2]:
# Exercise 7: Stock Information Database

# Create your database
stock_database = {
    "AAPL": ("Apple Inc.", "Technology", 150.75),
    "MSFT": ("Microsoft", "Technology", 320.50),
    "GOOGL": ("Google","Technology", 135.20),
    "JPM": ("JPMorgan Chase", "Finance", 145.30),
    "TSLA": ("Tesla", "Automotive", 255.10)
}

# Write your code here - Print tech stocks
print("Technology Stocks: ")
for name, info in stock_database.items():
    company, sector, price = info
    if sector == "Technology":
        print(name, company, price)
#name = AAPL, MSFT, EXC = the KEYS and what info is the values of the tuple.
#then below we are listing what is within that tuple info.

#find the most expensive stock
most_expensive_symbol = None
highest_price = 0

for symbol, info in stock_database.items():
    price = info[2]
    if price > highest_price:
        highest_price = price
        most_expensive_symbol = symbol
# it goes through the loop comparing the stocks one by one, by setting the price just used equal to the highest price in the next loop
print("Most expensive stock:", most_expensive_symbol, highest_price)

#calculate average price
total_price = 0
count = 0

for info in stock_database.values():
    total_price += info[2]
    #this line above is adding all stock values
    count += 1
    #this line above is a counter for how many times the loop runs through - it will be however many stocks there are

average_price = total_price/count
print("Average stock price:", average_price)



Technology Stocks: 
AAPL Apple Inc. 150.75
MSFT Microsoft 320.5
GOOGL Google 135.2
Most expensive stock: MSFT 320.5
Average stock price: 201.37


### Exercise 8: List Comprehension Practice

Given a list of stock prices, use list comprehensions to create:

**a)** A list of prices rounded to whole numbers

In [ ]:
# Exercise 8a: Round prices
prices = [150.75, 153.30, 149.50, 155.00, 157.25]

# Write your code here



**b)** A list of only prices above $152

In [ ]:
# Exercise 8b: Filter high prices
prices = [150.75, 153.30, 149.50, 155.00, 157.25]

# Write your code here



**c)** A list of percentage changes, where each price is compared to the first price

Formula: `((current - first) / first) * 100`

In [ ]:
# Exercise 8c: Calculate percentage changes
prices = [150.75, 153.30, 149.50, 155.00, 157.25]

# Write your code here



---

## Summary: Choosing the Right Data Structure

| Data Structure | Ordered | Mutable | Duplicates | Use Case |
|---------------|---------|---------|------------|----------|
| **List** | ✓ | ✓ | ✓ | Sequences, collections that change |
| **Tuple** | ✓ | ✗ | ✓ | Fixed data, function returns |
| **Dictionary** | ✗* | ✓ | Keys: ✗, Values: ✓ | Key-value mappings |
| **Set** | ✗ | ✓ | ✗ | Unique items, set operations |

*Python 3.7+ maintains insertion order for dictionaries

### Quick Reference:

**Lists:**
- Create: `my_list = [1, 2, 3]`
- Access: `my_list[0]`
- Methods: `.append()`, `.remove()`, `.sort()`, `.reverse()`

**Tuples:**
- Create: `my_tuple = (1, 2, 3)`
- Access: `my_tuple[0]`
- Unpack: `a, b, c = my_tuple`

**Dictionaries:**
- Create: `my_dict = {"key": "value"}`
- Access: `my_dict["key"]` or `my_dict.get("key")`
- Methods: `.keys()`, `.values()`, `.items()`

**Sets:**
- Create: `my_set = {1, 2, 3}`
- Methods: `.add()`, `.remove()`, `.union()`, `.intersection()`

**List Comprehensions:**
- Basic: `[x for x in iterable]`
- With condition: `[x for x in iterable if condition]`
- Transform: `[expression(x) for x in iterable]`

---

## Testing and Submission

**Before moving on:**
1. Test all exercises with different inputs
2. Understand when to use each data structure
3. Practice list comprehensions vs traditional loops
4. Have your instructor manually check your work

**Great work completing Lab 5!** 🎉